interfaces: allow access to the file locking for cryptosetup in the dm-crypt interface

bugraaydogar opened the pull request:
I created a snap that tries to use `cryptsetup` utility and installed it in `dev-mode`. According to the logs; it requires access to the  `/run/cryptosetup/` directory and also creates file(s) inside.

```
= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="open" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/" pid=720119 comm="cryptsetup" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
File: /run/cryptsetup/ (read)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="mknod" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="c" denied_mask="c" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="open" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="wrc" denied_mask="wrc" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="file_lock" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="wk" denied_mask="wk" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="unlink" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="d" denied_mask="d" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="open" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/" pid=720119 comm="cryptsetup" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
File: /run/cryptsetup/ (read)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="mknod" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="c" denied_mask="c" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="open" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="wrc" denied_mask="wrc" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="file_lock" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="wk" denied_mask="wk" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:03
Log: apparmor="ALLOWED" operation="unlink" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="d" denied_mask="d" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:10
Log: apparmor="ALLOWED" operation="open" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/" pid=720119 comm="cryptsetup" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
File: /run/cryptsetup/ (read)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:10
Log: apparmor="ALLOWED" operation="mknod" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="c" denied_mask="c" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:10
Log: apparmor="ALLOWED" operation="open" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="wrc" denied_mask="wrc" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:10
Log: apparmor="ALLOWED" operation="file_lock" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="wk" denied_mask="wk" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)


= AppArmor =
Time: May 11 11:55:10
Log: apparmor="ALLOWED" operation="open" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/" pid=720119 comm="cryptsetup" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
File: /run/cryptsetup/ (read)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:10
Log: apparmor="ALLOWED" operation="mknod" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="c" denied_mask="c" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:10
Log: apparmor="ALLOWED" operation="open" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="wrc" denied_mask="wrc" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

= AppArmor =
Time: May 11 11:55:10
Log: apparmor="ALLOWED" operation="file_lock" profile="snap.test-content-slot.slot-exec" name="/run/cryptsetup/L_8:17" pid=720119 comm="cryptsetup" requested_mask="wk" denied_mask="wk" fsuid=0 ouid=0
File: /run/cryptsetup/L_8:17 (write)
Suggestions:
* adjust program to use $SNAP_DATA
* adjust program to use /run/shm/snap.$SNAP_NAME.*
* adjust program to use /run/snap.$SNAP_NAME.*
* adjust snap to use snap layouts (https://forum.snapcraft.io/t/snap-layouts/7207)

```

mardy reviewed (approved):
LGTM, thanks!

bugraaydogar commented:
Hi @alexmurray,

Would it be possible to review this interface modification?

Thanks

alexmurray reviewed (commented):
The comment mentions file locking but the `k` (lock) permission is not included for the AppArmor rule - is the comment correct?

bugraaydogar commented:
Thanks, good catch. Just updated.

alexmurray commented on interfaces/builtin/dm_crypt.go:
```diff
@@ -61,6 +61,9 @@ mount options=(rw,nosuid,nodev) /dev/dm-[0-9]* -> /{,run/}media/**,
 /run/mount/utab* wrlk,
+
+# Allow access to the file locking mechanism
+/run/cryptsetup rwk,
```
Should this also include `/var/run/` as well as the more standard `/run` prefix? I notice we generally do this for other locations under `/run`. ie should it instead be:

```
/{,var/}run/cryptsetup rwk,
```

Also does this require access to any files within `/run/cryptsetup` or only just to the directory itself?

Finally, if we do change this to be `/{,var/}run/cryptsetup rwk,` then line 63 should also be changed to add the `{,var/}` prefix too.

bugraaydogar replied:
`/run/cryptsetup` is just a file, @mardy any feedback from your perspective?

alexmurray replied:
Are you sure? On my local machine this is a directory.

alexmurray reviewed (changes requested):
Thanks but I think this may also need to include support for the `/var/run` path as well as `/run` as I think this was used on older Ubuntu releases (at least we use the `/{,var/}run/` pattern in a lot of other interfaces).

alexmurray reviewed (changes requested):
I'm not very convinced by these changes - if you are not certain whether this is a directory or a file and then whether access to any of the files within the directory are required then I don't think this change is ready. Please make sure you understand exactly what files/directories are required and what permissions are required (since I only suggested adding the lock permission based on your comment but I have no actual knowledge if this is really required or not - I hope though as the person suggesting this change that you do know if this is actually required).

Perhaps it would be worth including (as comments in this PR) and associated AppArmor denials from dmesg so we can help understand exactly what issue you are trying to solve with this PR.

bugraaydogar commented:
Thanks for the feedback. I agree that I should have reproduced and added the logs before.

I updated the PR content and attached the snappy-debug logs. Please keep in mind that the snap is installed with `dev-mode`.
I only have some concerns about `c` and `d` access mode keywords since they are not documented in the AppArmor's man page. Nevertheless, I added them to the interface AppArmor rules. Any comments on these?

bugraaydogar commented:
`c` (create) and `d` (delete) keywords are both maps to `w` (write) policy. No need to explicitly add them in the AppArmor rules.

alexmurray reviewed (approved):
LGTM! Thanks for making all the changes along the way.

mvo5 reviewed (approved):
Looks fine to me but needs a +1 from security

bboozzoo reviewed (approved):
LGTM, thank you!
